src/gsi/hybrid_ensemble_isotropic.F90

```diff
@@ -3939,32 +4073,56 @@ subroutine ckgcov_a_en_new_factorization_ad(ig,z,a_en)
   endif
 
 ! Apply vertical smoother on each ensemble member
-  iadvance=1 ; iback=2
-!$omp parallel do schedule(static,1) private(k)
-  do k=1,n_ens
-
-     call new_factorization_rf_z(a_en(k)%r3(ipnt)%q,iadvance,iback,ig)
- 
-  enddo
-
+  if(regional.and.l_mgbf_loc) then
+     allocate(work_mgbf(obj_mgbf(ig)%km_a_all,obj_mgbf(ig)%nm,obj_mgbf(ig)%mm))
+     work_mgbf=zero
+     iadvance=1 ; iback=2
+!$omp parallel do schedule(dynamic,1) private(k,kk,ii,i,j,work_tmp)
+     do k=1,n_ens
+        if(.not.obj_mgbf(ig)%l_vertical_filter) call new_factorization_rf_z(a_en(k)%r3(ipnt)%q,iadvance,iback,ig)
+        do kk=1,grd_loc%nsig
```

Comment:
This kind of transfer/copying from line 4083  to 4095  is a repeated pattern , could all the similar loops be replaced a  by calling  a sub/ or function?

Reply:
I replaced these parts to a subroutine "map_work_mgbf".